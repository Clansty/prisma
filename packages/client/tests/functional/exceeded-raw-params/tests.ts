import crypto from 'crypto'

import testMatrix from './_matrix'
// @ts-ignore
import type { Prisma as PrismaNamespace, PrismaClient } from './node_modules/@prisma/client'

declare let prisma: PrismaClient
declare let Prisma: typeof PrismaNamespace

/**
 * Regression test for #9326
 * Prepared statement parameters exceeded
 */
testMatrix.setupTestSuite(
  ({ provider }) => {
    test('lots of raw parameters', async () => {
      let insertValues: PrismaNamespace.Sql

      if (provider === 'mysql') {
        insertValues = Prisma.sql`INSERT INTO \`Resource\` (\`id\`, \`field\`) VALUES ("${0}", ${0})`
      } else {
        insertValues = Prisma.sql`INSERT INTO "Resource" ("id", "field") VALUES ("${0}", ${0})`
      }

      for (let i = 1; i < 33000; i++) {
        insertValues = Prisma.join([insertValues, Prisma.sql`("${i}", ${0})`], ',')
      }

      const query = prisma.$executeRaw(insertValues)

      await expect(query).rejects.toMatchPrismaErrorSnapshot()
    })
  },
  {
    optOut: {
      from: ['mongodb', 'sqlite'],
      reason: '$executeRaw only works on SQL based providers, SQLite does not have raw parameters limits',
    },
  },
)
