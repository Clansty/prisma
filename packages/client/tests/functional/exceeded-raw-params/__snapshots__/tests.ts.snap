// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exceeded-raw-params (provider=cockroachdb) lots of raw parameters 1`] = `

Invalid \`prisma.$executeRaw()\` invocation:


Raw query failed. Code: \`N/A\`. Message: \`N/A\`
`;

exports[`exceeded-raw-params (provider=mysql) lots of raw parameters 1`] = `

Invalid \`prisma.$executeRaw()\` invocation:


Raw query failed. Code: \`N/A\`. Message: \`N/A\`
`;

exports[`exceeded-raw-params (provider=postgresql) lots of raw parameters 1`] = `

Invalid \`prisma.$executeRaw()\` invocation:


Raw query failed. Code: \`N/A\`. Message: \`N/A\`
`;

exports[`exceeded-raw-params (provider=sqlserver) lots of raw parameters 1`] = `

Invalid \`prisma.$executeRaw()\` invocation:


Raw query failed. Code: \`8003\`. Message: \`The incoming request has too many parameters. The server supports a maximum of 2100 parameters. Reduce the number of parameters and resend the request.\`
`;
